### PR TITLE
Do not fail with 'conan remove -r remote -p' if there are no packages in the remote

### LINKED ISSUE #7332: [bug] conan remove fails when using -p without package id together with -r
When I do `conan remove -f -p -r <remote> <existing_reference>` with a reference for which there is only a recipe but no binary package on the remote Conan fails with the following error message:

```
ERROR: {
  "errors" : [ {
    "status" : 404,
    "message" : "Couldn't find 'user/pkg/0.1.0/develop/d8cc7d7e1317dcbc71d65614521acffd/package'"
  } ]
}. [Remote: my-remote]
```

I would expect it to succeed even if there is nothing to do.

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: Bugfix: Do not fail for `conan remove -r remote -p` when there are no packages in the remote.
Docs: omit

Check there are no packages before trying to remove the package folder in the server. It prevents a 404 error.

close https://github.com/conan-io/conan/issues/7332

#REVISIONS: 1

### CODE REVIEW

**Review by czoido — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

File: conans/client/rest/rest_client_v1.py [hunk 1]
```diff
@@ -306,10 +307,13 @@ def _remove_conanfile_files(self, ref, files):
     @handle_return_deserializer()
     def remove_packages(self, ref, package_ids=None):
         """ Remove any packages specified by package_ids"""
-        self.check_credentials()
-        payload = {"package_ids": package_ids}
-        url = self.router.remove_packages(ref)
-        return self._post_json(url, payload)
+        pcks = self.search_packages(ref, query=None)
```
Comment (memsharded) [inline conans/client/rest/rest_client_v1.py]: I would only do the search if ``package_ids is None``, in other cases it should still raise.

Actually, I'd prefer capturing the raise, checking if package_ids is None, and re-raising if not, but I guess it is not easy to differentiate the 404 that the ``packages`` folder is missing or the 404 that the recipe is missing.

File: conans/client/rest/rest_client_v2.py [hunk 2]
```diff
@@ -246,13 +246,16 @@ def remove_packages(self, ref, package_ids=None):
...
             if not package_ids:
-                url = self.router.remove_all_packages(ref)
-                response = self.requester.delete(url, auth=self.auth, headers=self.custom_headers,
-                                                 verify=self.verify_ssl)
-                if response.status_code != 200:  # Error message is text
-                    # To be able to access ret.text (ret.content are bytes)
-                    response.charset = "utf-8"
-                    raise get_exception_from_error(response.status_code)(response.text)
+                # Check there are packages to remove
+                package_search_url = self.router.search_packages(ref)
```
Comment (memsharded) [inline conans/client/rest/rest_client_v2.py]: The ``search_packages`` is  quite slow, tbh, I am not very excited about this approach. I would say:

- This seems something to be fixed in the server side. I think it is a bug and can be fixed, it only belongs to the Conan API
- In the meantime, I would try to workaround it in the client side with a try - except 404 and only checking the ``search_packages()`` in that case. But annotating clearly it is a temporary workaround.

**Comment by jgsogo** [comment]
Ok, I'll use this other approach, but I don't think it is something to fix in the server: the API is REST-based and we request a DELETE on a resource that doesn't exist, `404` is expected.

**Comment by memsharded** [comment]
> Ok, I'll use this other approach, but I don't think it is something to fix in the server: the API is REST-based and we request a DELETE on a resource that doesn't exist, 404 is expected.

Yes you are right, it is true that in Api V2 it is a DELETE over ``ref/revision/packages`` path. It was not this way in Api V1.
So I agree. For me, the correct scenario would be that the 404 error is able to return the difference: it errors because the ``packages`` folder is not there, or it is failing because the package recipe is not there at all. There it would be possible for the client to act accordingly, raise RecipeNotFound or pass without raising. As this change requires too much work for little value, yes, please at least change the implementation to do the search() only for the failing 404 case.

**Review by memsharded — CHANGES_REQUESTED** [review]
Please change the impl so it only does ``search_packages`` for 404 responses, after the error, not before.

File: conans/client/rest/rest_client_v1.py [hunk 3]
```diff
@@ -309,7 +310,13 @@ def remove_packages(self, ref, package_ids=None):
         self.check_credentials()
         payload = {"package_ids": package_ids}
         url = self.router.remove_packages(ref)
-        return self._post_json(url, payload)
+        ret = self._post_json(url, payload)
+        if ret.status_code == 404:
```
Comment (memsharded) [inline conans/client/rest/rest_client_v1.py]:
```suggestion
        if package_ids is None and ret.status_code == 404:
```

**Review by memsharded — APPROVED** [review]
Check the suggestion first, then merge, looking good.

File: conans/client/rest/rest_client_v1.py [hunk 4]
```diff
@@ -311,7 +311,7 @@ def remove_packages(self, ref, package_ids=None):
         payload = {"package_ids": package_ids}
         url = self.router.remove_packages(ref)
         ret = self._post_json(url, payload)
-        if package_ids is None and ret.status_code == 404:
+        if package_ids is not None and ret.status_code == 404:
```
Comment (memsharded) [inline conans/client/rest/rest_client_v1.py]: I don't understand this change. The error in the server happens when we try to remove all the packages, and we don't specify ``package_ids``. If we explictly specify ``-p=<package_id>``, then, the 404 error is expected and desired.

**Comment by jgsogo** [comment]
* `package_ids` is always a list (I'm removing the default value)
 * I'm  capturing any exception from the double-check operation, it would be misleading if it propagates and it would hide the original one.

**Pull request merged**